ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,567 @@
+                EnumSet.of(StoreGetOptions.Store_Include_Deleted, StoreGetOptions.Store_Include_Expired));
+            if (originalPutInfo != null && !isExpired(originalPutInfo.getExpiresAtMs(), scanTimeWithOffsetInMs)) {
+              newScanResults.updateSegmentBaseMap(segmentName, originalPutInfo.getSize());
+              long removalTime = originalPutInfo.getExpiresAtMs() == Utils.Infinite_Time ?
+                  lastModifiedTime : Math.min(lastModifiedTime, originalPutInfo.getExpiresAtMs());
+              newScanResults.updateSegmentBucket(removalTime, segmentName, originalPutInfo.getSize() * -1);
+            }
+          }
+        } else {
+          String putSegmentName = putIndexValue.getOffset().getName();
+          if (lastModifiedTime < scanTimeWithOffsetInMs) {
+            processDeleteIndexForSegment(putIndexValue, newScanResults, putSegmentName, scanTimeWithOffsetInMs);
+          } else if (isWithinWindow(lastModifiedTime, scanWindowBoundary, segmentScanOffset)) {
+            // Delete records that we need to capture in buckets
+            newScanResults.updateSegmentBucket(lastModifiedTime, putSegmentName, putIndexValue.getSize() * -1);
+            if (newScanResults.checkIfBucketEntryExists(putIndexValue.getExpiresAtMs(), putSegmentName) &&
+                putIndexValue.getExpiresAtMs() >= lastModifiedTime) {
+              newScanResults.updateSegmentBucket(putIndexValue.getExpiresAtMs(), putSegmentName,
+                  putIndexValue.getSize());
+            }
+          }
+        }
+      } else {
+        // Put record that is not expired
+        newScanResults.updateSegmentBaseMap(segmentName, messageInfo.getSize());
+        if (messageInfo.getExpirationTimeInMs() != -1 &&
+            isWithinWindow(messageInfo.getExpirationTimeInMs(), scanWindowBoundary, segmentScanOffset)) {
+          newScanResults.updateSegmentBucket(messageInfo.getExpirationTimeInMs(), segmentName,
+              messageInfo.getSize() * -1);
+        }
+      }
+    }
+
+    /**
+     * Process a record in the index to collect data for valid data size per container.
+     * purpose.
+     * @param messageInfo The {@link MessageInfo} to be processed.
+     * @param backwardSearchSpan A {@link FileSpan} that spans from the beginning of the index to the end of the
+     *                           previous index segment.
+     * @throws StoreException
+     */
+    private void populateValidSizePerContainer(MessageInfo messageInfo, FileSpan backwardSearchSpan) throws StoreException {
+      if (isExpired(messageInfo.getExpirationTimeInMs(), scanTimeInMs)) {
+        return;
+      }
+      if (messageInfo.isDeleted()) {
+        // delete record (may need to find and remove previously counted put data size).
+        IndexValue putIndexValue = index.findKey(messageInfo.getStoreKey(), backwardSearchSpan);
+        if (putIndexValue == null) {
+          return;
```

Comment:
At the moment there is only time stamp at the index segment level and not at the IndexValue level. The implementation here will be different once #541 is merged. Without #541 the best you can do is either assume the put occurred at the last modified time of previous index segment and the delete occurred at the last modified time of current index segment (which could be very wrong). Or, you treat the put and delete occurred at the same time (what's currently implemented).